cas-server-support-osf/src/main/java/io/cos/cas/web/flow/OpenScienceFrameworkTerminateSessionAction.java

```diff
@@ -0,0 +1,151 @@
+    private final CookieRetrievingCookieGenerator warnCookieGenerator;
+    /**
+     * Creates a new instance with the given parameters.
+     * @param cas Core business logic object.
+     * @param tgtCookieGenerator TGT cookie generator.
+     * @param warnCookieGenerator Warn cookie generator.
+     */
+    public OpenScienceFrameworkTerminateSessionAction(
+            final CentralAuthenticationService cas,
+            final CookieRetrievingCookieGenerator tgtCookieGenerator,
+            final CookieRetrievingCookieGenerator warnCookieGenerator) {
+        this.centralAuthenticationService = cas;
+        this.ticketGrantingTicketCookieGenerator = tgtCookieGenerator;
+        this.warnCookieGenerator = warnCookieGenerator;
+    }
+
+    /**
+     * The Terminate Action.
+     * @param context The Request Context
+     * @return the Event success or finish
+     */
+    public Event terminate(final RequestContext context) {
+        // in login's webflow : we can get the value from context as it has already been stored
+        String tgtId = WebUtils.getTicketGrantingTicketId(context);
+
+        String institutionId = "osf";
+        Boolean remotePrincipal = Boolean.FALSE;
+
+        // for logout, we need to get the cookie's value
+        if (tgtId == null) {
+            final HttpServletRequest request = WebUtils.getHttpServletRequest(context);
+            tgtId = this.ticketGrantingTicketCookieGenerator.retrieveCookieValue(request);
+        }
+        // for institution logout, we need to get the TGT which contains the institutionId
+        if (tgtId != null) {
+            TicketGrantingTicket tgt = null;
+            try {
+                tgt  = centralAuthenticationService.getTicket(tgtId, TicketGrantingTicket.class);
+            }
+            catch (final Exception e) {
+                //  ignore
+            }
+            if (tgt != null) {
+                final Authentication auth = tgt.getAuthentication();
+                if (auth != null) {
+                    institutionId = (String) auth.getAttributes().get("institutionId");
+                    remotePrincipal = (Boolean) auth.getAttributes().get("remotePrincipal");
+                }
+            }
+
+            WebUtils.putLogoutRequests(context, this.centralAuthenticationService.destroyTicketGrantingTicket(tgtId));
+        }
+        final HttpServletResponse response = WebUtils.getHttpServletResponse(context);
+        this.ticketGrantingTicketCookieGenerator.removeCookie(response);
+        this.warnCookieGenerator.removeCookie(response);
+
+        final HttpServletRequest request = WebUtils.getHttpServletRequest(context);
+        final Cookie[] cookies = request.getCookies();
+        if (cookies != null) {
+            for (final Cookie cookie : cookies) {
+                if (cookie.getName().startsWith(
+                        OpenScienceFrameworkPrincipalFromRequestRemoteUserNonInteractiveCredentialsAction.SHIBBOLETH_COOKIE_PREFIX)) {
```

Comment:
we do not need to do all this work with the cookie we should redirect to the logout_url from the database/institution. e.g. `https://accounts.osf.io/Shibboleth.sso/Logout?return=https://shibboleth.nyu.edu/idp/profile/Logout` - let shibboleth handle deleting its own cookie and perform the final redirect (return parameter)